src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +661,210 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+            {
+                // Add to the list of leaves to prune
+                leavesToPrune.push_back((*lpIter)->parent);
+            }
+            // No else, does not meet the pruning criteria
+
+            // Erase the actual motion, by first freeing the state
+            si_->freeState((*lpIter)->state);
+
+            // then deleting the pointer
+            delete *lpIter;
+
+            // And finally removing it from the list, erase returns the next iterator
+            lpIter = leavesToPrune.erase(lpIter);
+
+            // Update our counter
+            ++numPruned;
+        }
+
+        // Update the cost at which we've pruned:
+        prunedCost_ = pruneTreeCost;
+
+        // And if we're using informed sampling, the measure to which we've pruned
+        if (useInformedSampling_)
+        {
+            prunedInfMeasure_ = infSampler_->getInformedMeasure(prunedCost_);
         }
     }
 
-    // To create the new nn takes one order of magnitude in time more than just checking how many
-    // states would be pruned. Therefore, only prune if it removes a significant amount of states.
-    if ((double)pruneScratchSpace_.newTree.size() / tree_size < pruneStatesThreshold_)
-    {
-        for (std::size_t i = 0; i < pruneScratchSpace_.toBePruned.size(); ++i)
-            deleteBranch(pruneScratchSpace_.toBePruned[i]);
+    return numPruned;
+}
 
-        nn_->clear();
-        nn_->add(pruneScratchSpace_.newTree);
 
-        return (tree_size - pruneScratchSpace_.newTree.size());
+bool ompl::geometric::RRTstar::leafPruningCondition(Motion *motion, const base::Cost& pruneTreeCost)
+{
+    // Variable
+    // Whether the motion is a leaf to be pruned. Be pessimistic
+    bool pruneThisLeaf = false;
+
+    // Is it a leaf?
+    if (motion->children.empty() == true)
+    {
+        // Is it above the pruning criterion?
+        if (opt_->isCostBetterThan(pruneTreeCost, solutionHeuristic(motion)))
+        {
+            // Mark as prunable
+            pruneThisLeaf = true;
+        }
+        // No else, it does not meet the pruning condition.
     }
-    return 0;
+    // No else, it has children
+
+    return pruneThisLeaf;
+}
+
+ompl::base::Cost ompl::geometric::RRTstar::solutionHeuristic(const Motion *motion, const bool estimate) const
+{
+    base::Cost costToCome;
+    if (estimate)
+        costToCome = opt_->motionCost(startMotion_->state, motion->state); // lower-bounding cost from the start to the state
+    else
+        costToCome = motion->cost; // current cost from the state to the goal
+
+    const base::Cost costToGo = opt_->costToGo(motion->state, pdef_->getGoal().get()); // lower-bounding cost from the state to the goal
+    return opt_->combineCosts(costToCome, costToGo); // add the two costs
 }
 
-void ompl::geometric::RRTstar::deleteBranch(Motion *motion)
+void ompl::geometric::RRTstar::setSampleRejection(const bool reject)
```

Comment:
If we set sample rejection after a sampler has been allocated, we need to create a new sampler.
If a sampler has not yet been allocated, we don't need to.